### PR TITLE
Difficulty length check

### PR DESCRIPTION
**Pull request by AionJayT**
## Notice

It is not allowed to submit your PR to the master branch directly, please submit your PR to the master-pre-merge branch.

## Description

Please include a brief summary of the change that this pull request proposes. Include any relevant motivation and context. List any dependencies required for this change.

- The PR is mainly for adding difficulty length check in the header rule to prevent adversary padding big 
date into this field.

Fixes Issue # .

## Type of change

Insert **x** into the following checkboxes to confirm (eg. [x]):
- [x] Bug fix.
- [ ] New feature.
- [ ] Enhancement.
- [x] Unit test.
- [ ] Breaking change (a fix or feature that causes existing functionality to not work as expected).
- [ ] Requires documentation update.

## Testing

Please describe the tests you used to validate this pull request. Provide any relevant details for test configurations as well as any instructions to reproduce these results.

-

## Verification

Insert **x** into the following checkboxes to confirm (eg. [x]):
- [x] I have self-reviewed my own code and conformed to the style guidelines of this project.
- [x] New and existing tests pass locally with my changes.
- [x] I have added tests for my fix or feature.
- [ ] I have made appropriate changes to the corresponding documentation.
- [x] My code generates no new warnings.
- [ ] Any dependent changes have been made.

### CODE REVIEW

File: modMcf/src/org/aion/mcf/types/AbstractBlockHeader.java [hunk 1]
```diff
@@ -24,13 +24,17 @@
 
 import java.math.BigInteger;
 import org.aion.base.type.Address;
+import org.aion.log.AionLoggerFactory;
+import org.aion.log.LogEnum;
+import org.slf4j.Logger;
```
Comment (AlexandraRoatis) [inline modMcf/src/org/aion/mcf/types/AbstractBlockHeader.java]: some unused imports

File: modMcf/src/org/aion/mcf/types/AbstractBlockHeader.java [hunk 2]
```diff
@@ -24,13 +24,17 @@
...
 import org.spongycastle.util.BigIntegers;
 
 /** Abstract BlockHeader. */
 public abstract class AbstractBlockHeader {
 
     public static final int NONCE_LENGTH = 32;
     public static final int SOLUTIONSIZE = 1408;
+    static final int DIFFICULTY_LENGTH = 16;
```
Comment (AlexandraRoatis) [inline modMcf/src/org/aion/mcf/types/AbstractBlockHeader.java]: can be private

File: modMcf/test/org/aion/valid/DifficultyRuleTest.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+                        new byte[NONCE_LENGTH],
+                        new byte[SOLUTIONSIZE]);
+    }
+
+    /**
+     * Checks if the {@link AionDifficultyRule#validate} returns {@code false} when the difficulty
+     * data is <b>greater</b> than the difficulty data length.
+     */
+    @Test
+    public void testInvalidDiffcultyLength() {
```
Comment (AlexandraRoatis) [inline modMcf/test/org/aion/valid/DifficultyRuleTest.java]: minor misspelling of "difficulty"

File: modMcf/test/org/aion/valid/DifficultyRuleTest.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        // test output
+        assertThat(actual).isFalse();
+    }
+
+    /**
+     * Checks if the {@link AionDifficultyRule#validate} returns {@code true} when the difficulty
+     * data is <b>equal</b> the difficulty data length.
+     */
+    @Test
+    public void testDiffcultyLength() {
```
Comment (AlexandraRoatis) [inline modMcf/test/org/aion/valid/DifficultyRuleTest.java]: same misspeling

File: modMcf/src/org/aion/mcf/types/AbstractBlockHeader.java (hunk 2 above)
Comment (aion-kelvin) [inline modMcf/src/org/aion/mcf/types/AbstractBlockHeader.java]: optional: I think "MAX_DIFFICULTY_LENGTH" would be a more clear name

File: modMcf/src/org/aion/mcf/types/AbstractBlockHeader.java [hunk 5]
```diff
@@ -166,6 +170,10 @@ public void setTransactionsRoot(byte[] stateRoot) {
     }
 
     public BigInteger getDifficultyBI() {
+        if (difficulty == null || difficulty.length > DIFFICULTY_LENGTH) {
+            AionLoggerFactory.getLogger("CONS").error("Invalid difficulty length!");
+            return BigInteger.ZERO;
```
Comment (aion-kelvin) [inline modMcf/src/org/aion/mcf/types/AbstractBlockHeader.java]: Can we add a JavaDoc to this that explains that BigInteger.ZERO is used if difficulty length is invalid?  And also mention which range of numbers are valid (I'm guessing it's supposed to be the positive integers, but would be good to make it very clear)

**Pull request merged**